### DISCUSSION TITLE
Sharing data between sibling components

### DISCUSSION THREAD

**Discussion by Atulin** [lit/lit#2927] [Q&A]
I'm currently working on an SSR project only enhanced with components here and there, that's why I can't use a parent component to hold the state - there is none.

I need to add between 0 and an arbitrary number of a certain component to the page. That component gets its initial state from an API, but said state is the same for each instance of it.

Thus, it wouldn't make sense to call the API 200 times, once per component, if I can *somehow* call it only once and cache the result for all of the components.

Part of me wants to just do something like

```js
export const store = {
    data: null
}
```

and then `import` it to every component, and set the component's state taking this cache into account, something like

```js
if (store.data === null) {
    this.thing = store.data = await getData();
} else {
    this.thing = store.data;
}
```

but it seems a very naive approach. Not to mention, I'm not sure how much of a guarantee I have of only one component actually fetching that data.

Another approach would be maybe loading that data in the store itself, not in the components. Something like

```ts
class Store {
    data: Data;
    constructor(data: Data) {
        this.data = data;
    } 
    async init() {
        return new Store(await getData());
    } 
}

const store = await Store.init();
export default store;
```

But I'm not sure how good of an idea this one is either.

All results of googling tell me to use something like Redux and what not, but I really don't need this much added complexity and this many features.

Any recommended ways of going about it?

(also, I wanted to ask on Slack, but all invite links are dead)

**Comment by dbwodlf3** [comment]
How about use customEvent?

https://lit.dev/docs/composition/component-composition/#passing-data-up-and-down-the-tree
https://lit.dev/docs/composition/component-composition/#passing-data-across-the-tree

```ts
//
window.addEventListener("FetchData", (e)=>{
    this.thing = e.detail.store.data 
    //...
})
//.... 
getData().then(()=>{
  window.dispatchEvent( new CustomEvent("FetchData"));
})
```

**Comment by kevinpschaaf** [comment]
Since you mention there is no parent component, this suggestion seems perfectly reasonable for its simplicity:

> Another approach would be maybe loading that data in the store itself, not in the components.

To decouple the data fetching from the components themselves, @dbwodlf3's answer seems reasonable. And as a next step up in sophistication, you could make wrapper component (in the light DOM) which is responsible for doing the fetching, and use events in a similar way, aka:

```
<data-provider>
  <data-consumer></data-consumer>
  <data-consumer></data-consumer>
</data-provider>
```

You could implement this bespoke, by just having `<data-provider>` do `addEventListener('request-data', () => /* store e.detail.callback */)` to listen for children who want its data and `<data-consumer>` do `this.dispatchEvent(new CustomEvent('request-data', {detail: {callback: (data) => this.data = data;}))` to ask for data. Then once `<data-provider>` has the data, it can call all the callbacks with it.

And if that started to feel too complex, you could look into [`@lit-labs/context`](https://github.com/lit/lit/tree/main/packages/labs/context#lit-labscontext) which wraps this pattern up in an easy to use API.
